json/json.go

```diff
@@ -0,0 +1,126 @@
+package json
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/sourcegraph/zoekt"
+	"github.com/sourcegraph/zoekt/query"
+)
+
+// defaultTimeout is the maximum amount of time a search request should
+// take. This is the same default used by Sourcegraph.
+const defaultTimeout = 20 * time.Second
+
+func JSONServer(searcher zoekt.Searcher) http.Handler {
+	s := jsonSearcher{searcher}
+	mux := http.NewServeMux()
+	mux.HandleFunc("/search", s.jsonSearch)
+	mux.HandleFunc("/list", s.jsonList)
+	return mux
+}
+
+type jsonSearcher struct {
+	Searcher zoekt.Searcher
+}
+
+type jsonSearchArgs struct {
+	Q    string
+	Opts *zoekt.SearchOptions
+}
+
+type jsonSearchReply struct {
+	Result *zoekt.SearchResult
+}
+
+type jsonListArgs struct {
+	Q    string
+	Opts *zoekt.ListOptions
+}
+
+type jsonListReply struct {
+	List *zoekt.RepoList
+}
+
+func (s *jsonSearcher) jsonSearch(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	w.Header().Add("Content-Type", "application/json")
+
+	if req.Method != "POST" {
+		jsonError(w, http.StatusMethodNotAllowed, "Only POST is supported")
+		return
+	}
+
+	searchArgs := jsonSearchArgs{}
+	err := json.NewDecoder(req.Body).Decode(&searchArgs)
+	if err != nil {
+		jsonError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+	if searchArgs.Q == "" {
+		jsonError(w, http.StatusBadRequest, "missing query")
+		return
+	}
+	if searchArgs.Opts == nil {
+		searchArgs.Opts = &zoekt.SearchOptions{}
```

Comment:
minor: in the webserver handler it does some adjustments to SearchOptions based on the number of results it expects to find which is pretty neat. This may be useful behaviour to have in this handler if opts is not set: https://github.com/sourcegraph/zoekt/blob/d1964a317f841afecc71b7210fb6ef4c3de26733/web/server.go#L290-L316

Reply:
Yup RPC does not do this, so on sourcegraph's side we implement that behaviour. So doing it will improve the ergonomics for the client of this API.

Reply:
I extracted a shared function that's used by both the new JSON endpoint and the webserver.  I put it in the new json package, but if you think there's a better place, let me know.